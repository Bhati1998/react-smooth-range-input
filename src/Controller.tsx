import * as React from 'react';
import { Animate } from 'react-simple-animate';
import FlipNumbers from 'react-flip-numbers';
import { bubbleWithTail, bubble } from './constants/svgPath';

const commonAnimationProps = {
  easeType: 'cubic-bezier(0.86, 0, 0.07, 1)',
  startStyle: {
    transform: 'translateY(0)',
  },
};

interface Props {
  onFocus: () => void;
  onBlur: () => void;
  buttonSize: number;
  height: number;
  value: number;
  dragX: number;
  showBubble: boolean;
  isControlByKeyBoard: boolean;
  isThin: boolean;
  onMouseDown: (MouseEvent) => void;
  onInteractEnd: (Event) => void;
  backgroundColor?: string;
  textBackgroundColor?: string;
  textColor?: string;
  isTouchDevice: boolean;
}

const flipNumberProps = {
  height: 14,
  width: 9,
  color: 'black',
  background: 'white',
  perspective: 370,
  durationSeconds: 0.4,
  play: true,
  numberStyle: { outline: '1px solid transparent' },
};

export default function Controller({
  onFocus,
  onBlur,
  buttonSize,
  height,
  dragX,
  showBubble,
  isControlByKeyBoard,
  value,
  onMouseDown,
  onInteractEnd,
  backgroundColor,
  textBackgroundColor,
  textColor,
  isThin,
  isTouchDevice,
}: Props) {
  return (
    <div
      tabIndex={0}
      {...(isTouchDevice
        ? {}
        : {
            onClick: e => e.stopPropagation(),
            onFocus: onFocus,
            onBlur: onBlur,
            onMouseDown: onMouseDown,
            onMouseUp: onInteractEnd,
          })}
      style={{
        top: `${(height - buttonSize) / 2}px`,
        position: 'absolute',
        cursor: '-webkit-grab',
        fontWeight: 600,
        transform: `translateX(${dragX}px)`,
        transition: isControlByKeyBoard ? '0.15s all ease-in' : '0s all',
      }}
      role="slider"
      aria-valuenow={value}
      aria-valuemin={2}
      aria-valuemax={30}
      aria-valuetext="Years of loan"
    >
      <input
        type="range"
        defaultValue={value.toString()}
        style={{
          visibility: 'hidden',
          width: 0,
          height: 0,
          display: 'block',
          position: 'absolute',
        }}
      />
      <Animate
        play={showBubble}
        {...commonAnimationProps}
        reverseDurationSeconds={0.3}
        durationSeconds={0.2}
        startStyle={{
          transform: 'translateY(0)',
        }}
        endStyle={{
          transform: `translateY(-${isThin ? 32 : 22}px) scale(1.65)`,
        }}
        easeType="cubic-bezier(0.86, 0, 0.07, 1)"
        render={({ style }) => (
          <svg
            style={{
              position: 'absolute',
              top: '-25px',
              left: 0,
              ...style,
            }}
            x="0px"
            y="0px"
            width={`${buttonSize}px`}
            height="64px"
            viewBox="0 0 40 64"
          >
            <path
              style={{
                transition: '0.3s all',
                fill: backgroundColor,
              }}
              d={showBubble ? bubbleWithTail : bubble}
            />
          </svg>
        )}
      />

      <Animate
        play={showBubble}
        {...commonAnimationProps}
        endStyle={{
          transform: `translateY(-${isThin ? 58 : 48}px) scale(1.3)`,
        }}
        durationSeconds={0.3}
        reverseDurationSeconds={0.1}
        render={({ style }) => (
          <div
            style={{
              background: textBackgroundColor,
              height: `${buttonSize}px`,
              width: `${buttonSize}px`,
              borderRadius: '50%',
              position: 'absolute',
              ...style,
            }}
          >
            <span
              style={{
                position: 'absolute',
                top: '10px',
                left: 0,
                width: `${buttonSize}px`,
                color: textColor,
              }}
            >
              <FlipNumbers {...flipNumberProps} numbers={value.toString()} />
            </span>
          </div>
        )}
      />
    </div>
  );
}
