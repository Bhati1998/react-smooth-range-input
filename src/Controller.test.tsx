import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Controller from './Controller';

describe('Controller', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <Controller
        buttonSize={20}
        dragX={2}
        height={20}
        isControlByKeyBoard={false}
        isThin={false}
        onBlur={() => {}}
        onFocus={() => {}}
        onInteractEnd={() => {}}
        onMouseDown={() => {}}
        showBubble={false}
        value={20}
      />,
    );
    expect(tree).toMatchSnapshot();
  });
});
