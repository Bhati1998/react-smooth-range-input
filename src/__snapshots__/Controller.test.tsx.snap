// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller should render correct when isThin set to true 1`] = `
<div
  aria-valuemax={30}
  aria-valuemin={0}
  aria-valuenow={20}
  aria-valuetext="20"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  role="slider"
  style={
    Object {
      "cursor": "-webkit-grab",
      "fontWeight": 600,
      "position": "absolute",
      "top": "5px",
      "transform": "translateX(2px)",
      "transition": "0s all",
    }
  }
  tabIndex={0}
>
  <svg
    height="64px"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": "-25px",
        "transform": "translateY(0)",
        "transition": "all 0.3s cubic-bezier(0.86, 0, 0.07, 1) 0s",
      }
    }
    viewBox="0 0 40 64"
    width="10px"
    x="0px"
    y="0px"
  >
    <path
      d="M38.8,43.9c0,10.4-8.5,18.9-18.9,18.9S1,54.4,1,43.9S9.5,24,19.9,24S38.8,33.5,38.8,43.9z"
      style={
        Object {
          "fill": undefined,
          "transition": "0.3s all",
        }
      }
    />
  </svg>
  <div
    style={
      Object {
        "background": undefined,
        "borderRadius": "50%",
        "height": "10px",
        "position": "absolute",
        "transform": "translateY(0)",
        "transition": "all 0.1s cubic-bezier(0.86, 0, 0.07, 1) 0s",
        "width": "10px",
      }
    }
  />
</div>
`;

exports[`Controller should render correctly 1`] = `
<div
  aria-valuemax={30}
  aria-valuemin={0}
  aria-valuenow={20}
  aria-valuetext="20"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  role="slider"
  style={
    Object {
      "cursor": "-webkit-grab",
      "fontWeight": 600,
      "position": "absolute",
      "top": "5px",
      "transform": "translateX(2px)",
      "transition": "0s all",
    }
  }
  tabIndex={0}
>
  <svg
    height="64px"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "top": "-25px",
        "transform": "translateY(0)",
        "transition": "all 0.3s cubic-bezier(0.86, 0, 0.07, 1) 0s",
      }
    }
    viewBox="0 0 40 64"
    width="10px"
    x="0px"
    y="0px"
  >
    <path
      d="M38.8,43.9c0,10.4-8.5,18.9-18.9,18.9S1,54.4,1,43.9S9.5,24,19.9,24S38.8,33.5,38.8,43.9z"
      style={
        Object {
          "fill": undefined,
          "transition": "0.3s all",
        }
      }
    />
  </svg>
  <div
    style={
      Object {
        "background": undefined,
        "borderRadius": "50%",
        "height": "10px",
        "position": "absolute",
        "transform": "translateY(0)",
        "transition": "all 0.1s cubic-bezier(0.86, 0, 0.07, 1) 0s",
        "width": "10px",
      }
    }
  />
</div>
`;
